litmus-portal/graphql-server/pkg/chaos-workflow/handler/handler.go

```diff
@@ -115,37 +116,191 @@ func UpdateWorkflow(ctx context.Context, input *model.ChaosWorkFlowInput, r *sto
+
+		// Filtering based on workflow name
+		if input.Filter.WorkflowName != nil {
+			var filteredResult []*model.WorkflowRun
+			for _, wfRun := range result {
+				if strings.Contains(wfRun.WorkflowName, *input.Filter.WorkflowName) {
+					filteredResult = append(filteredResult, wfRun)
+				}
+			}
+			result = filteredResult
+		}
+
+		// Filtering based on cluster name
+		if input.Filter.ClusterName != nil && *input.Filter.ClusterName != "All" {
+			var filteredResult []*model.WorkflowRun
+			for _, wfRun := range result {
+				if wfRun.ClusterName == *input.Filter.ClusterName {
+					filteredResult = append(filteredResult, wfRun)
+				}
+			}
+			result = filteredResult
+		}
+
+		// Filtering based on phase
+		if input.Filter.WorkflowStatus != nil && *input.Filter.WorkflowStatus != "All" {
+			var filteredResult []*model.WorkflowRun
+			for _, wfRun := range result {
+				if wfRun.Phase == string(*input.Filter.WorkflowStatus) {
+					filteredResult = append(filteredResult, wfRun)
+				}
+			}
+			result = filteredResult
+		}
+
+		// Filtering based on date range
+		if input.Filter.DateRange != nil {
+			var filteredResult []*model.WorkflowRun
+			for _, wfRun := range result {
+				if wfRun.LastUpdated >= input.Filter.DateRange.StartDate &&
+					wfRun.LastUpdated <= input.Filter.DateRange.EndDate {
+					filteredResult = append(filteredResult, wfRun)
+				}
+			}
+			result = filteredResult
+		}
+	}
+
+	switch {
+	case input.Sort != nil && input.Sort.Field == model.WorkflowRunSortingFieldTime:
+		// Sorting based on LastUpdated time
+		if input.Sort.Descending != nil && *input.Sort.Descending {
+			sort.SliceStable(result, func(i, j int) bool {
+				return result[i].LastUpdated > result[j].LastUpdated
+			})
+		} else {
+			sort.SliceStable(result, func(i, j int) bool {
+				return result[i].LastUpdated < result[j].LastUpdated
+			})
+		}
+	case input.Sort != nil && input.Sort.Field == model.WorkflowRunSortingFieldName:
+		// Sorting based on WorkflowName time
+		if input.Sort.Descending != nil && *input.Sort.Descending {
+			sort.SliceStable(result, func(i, j int) bool {
+				return result[i].WorkflowName > result[j].WorkflowName
+			})
+		} else {
+			sort.SliceStable(result, func(i, j int) bool {
+				return result[i].WorkflowName < result[j].WorkflowName
+			})
+		}
+	default:
+		// Default sorting: sorts it by LastUpdated time in descending order
+		sort.SliceStable(result, func(i, j int) bool {
+			return result[i].LastUpdated > result[j].LastUpdated
+		})
+	}
+
+	// Calculate length of result after filtering
+	totalNoOfRuns := len(result)
+
+	if input.Pagination != nil {
+		startIndex := input.Pagination.Page * input.Pagination.Limit
+		endIndex := (input.Pagination.Page + 1) * input.Pagination.Limit
+		if endIndex > len(result) {
+			endIndex = len(result)
+		}
+		result = result[startIndex:endIndex]
+	}
+
+	output := model.GetWorkflowsOutput{
+		TotalNoOfWorkflowRuns: totalNoOfRuns,
+		WorkflowRuns:          result,
+	}
```

Comment:
Filtering not optimised will be a very slow query. All the filters should be at db level.if we are manually doing thee filters there is no benefit of moving it to the backend

Reply:
Due to the current structure of DB, we were bind to go with this approach. To shift all operations at DB level, we will need another schema consisting of only workflow runs having workflow_id as a foreign key. But due to some reasons it was decided to not split the schemas and go with this approach.
The benefit of moving it to the backend is that it will reduce the network load. Previously all these operations were done in frontend which was even more unoptimized from an end user perspective.

Reply:
Shifting to backend doesn't mean it will make things more efficient, in a setup where there is only 1 gql server (current scenario) fo reach request serrver has to do so much computation(searching o(n)+ sorting o(nlogn) or possibly o(n^2)) and if there are 100 users querying for 100 workflows each the amount of time taken to process each request increases this locks up server resources. In a small scale setup this won't be visible but when there is more load/workflows it will take a toll. Either we make this more optimized or wait until the server can be scaled.

Reply:
I already told you about the current scenario. Do you have any better approach to this problem than the one that we have used? How can we make it more optimized?

Reply:
The current scenario will not be solved by this solution, I can assure you that, slower queries won't be any better that more network load. 
Will have to look into this later, currently have other urgent tasks. Better to postpone this, there is already a task going on to scale the server if by then you can't figure any optimised solution we can use it. Also if dividing into 2 separate collections makes this easier you can discuss that approach too, I personally don't see any problem there.

Reply:
I think a few days back, we have calculated the DB calls after dividing them into two collections. It will take more DB calls than the current implementation.

Reply:
The other benefit of this approach is the computation done in the frontend will be removed. I agree on that part that it will definitely slow the queries/response time.

Reply:
Try to see a join or Aggregated mongo query for this.

Reply:
The best approach according to me which we gave initially was to split it into different collections. I too want to shift it to the db layer but the current schema is not suitable for that. I'll have a solution in mind...will do some research and try that.
@gdsoumya @rajdas98

Reply:
Currently, I don't see any lag in the frontend for the browse workflow section even when we have a considerable amount of workflows. With further increase in workflow count it might start lagging but the effect on the sever will be more prominent with this solution. if the server starts to slow down for each query being made more resources are being held up that would mean less number of requests can be served. It not just slows the current request but also can lead to a compounding effect on all other requests. 
Also if we shift the searching/sorting to the server that would mean the frontend will make a network call/request to the server anytime a user wants to filter the workflows. Previously it was done on the frontend so it could possibly reuse the existing data from the previous call itself. This would also mean now we have more network load, specially because we are providing realtime search results - every time a user types a letter we filter the results, if this uses the backend that would mean each key press is a network call

Reply:
This problem needs further discussion and proper design if we want to do this right.